packages/arb-rpc-node/rpc/lockout.go

```diff
@@ -0,0 +1,338 @@
+	return newBatcher, nil
+}
+
+const ACCEPTABLE_SEQ_NUM_GAP int64 = 0
+
+func (b *LockoutBatcher) lockoutManager(ctx context.Context) {
+	holdingMutex := false
+	defer (func() {
+		if !holdingMutex {
+			b.mutex.Lock()
+		}
+		b.currentBatcher = &errorBatcher{
+			err: errors.New("sequencer lockout manager shutting down"),
+		}
+		b.mutex.Unlock()
+		holdingMutex = false
+		backgroundContext := context.Background()
+		b.redis.releaseLockout(backgroundContext, &b.lockoutExpiresAt)
+		b.redis.releaseLiveliness(backgroundContext, b.rpcURL, &b.livelinessExpiresAt)
+		logger.Debug().Msg("shut down sequencer lockout manager and released locks")
+		select {
+		case <-ctx.Done():
+			break
+		default:
+			// We aren't shutting down but the lockout manager died
+			logger.Error().Msg("sequencer lockout manager died but context isn't shutting down")
+			go (func() {
+				// We need this goroutine to die so the panic is printed out,
+				// but we also want to exit the process as a whole afterwards.
+				time.Sleep(time.Second)
+				b.errChan <- errors.New("sequencer lockout manager died")
+			})()
+		}
+	})()
+	for {
+		alive := true
+		if !b.hasSequencerLockout() {
+			currentSeqNum, err := b.core.GetMessageCount()
+			if err != nil {
+				logger.Warn().Err(err).Msg("error getting sequence number")
+				time.Sleep(5 * time.Second)
+				continue
+			}
+			if b.lastLockedSeqNum == nil || new(big.Int).Add(currentSeqNum, big.NewInt(ACCEPTABLE_SEQ_NUM_GAP)).Cmp(b.lastLockedSeqNum) < 0 {
+				alive = false
+				if b.livelinessExpiresAt.After(time.Now()) {
+					logger.Warn().Str("ourSeqNum", currentSeqNum.String()).Str("targetSeqNum", b.lastLockedSeqNum.String()).Msg("fell behind sequencer position")
+					b.redis.releaseLiveliness(ctx, b.rpcURL, &b.livelinessExpiresAt)
+				}
+			}
+			b.lastLockedSeqNum = b.redis.getLatestSeqNum(ctx)
+		}
+		if alive {
+			b.redis.acquireOrUpdateLiveliness(ctx, b.rpcURL, b.livelinessTimeout, &b.livelinessExpiresAt)
+			if b.livelinessExpiresAt.Before(time.Now()) {
+				logger.Warn().Str("rpc", b.rpcURL).Msg("failed to acquire liveliness lockout, is another sequencer running with this RPC URL?")
+			}
+		}
+		selectedSeq := b.redis.selectSequencer(ctx)
+		if selectedSeq == b.rpcURL {
+			if !holdingMutex {
+				b.mutex.Lock()
+				holdingMutex = true
+			}
+			if b.livelinessExpiresAt.After(time.Now()) {
+				b.redis.acquireOrUpdateLockout(ctx, b.rpcURL, b.lockoutTimeout, &b.lockoutExpiresAt)
+			}
+			if b.hasSequencerLockout() {
+				if b.currentBatcher != b.sequencerBatcher {
+					logger.Info().Str("rpc", b.rpcURL).Msg("acquired sequencer lockout")
+					targetSeqNum := b.redis.getLatestSeqNum(ctx)
+					b.lastLockedSeqNum = targetSeqNum
+					attemptCatchupUntil := b.lockoutExpiresAt.Add(time.Second * -10)
+					for {
+						currentSeqNum, err := b.core.GetMessageCount()
+						if err != nil {
+							logger.Warn().Err(err).Msg("error getting sequence number")
+							time.Sleep(5 * time.Second)
+							continue
```

Comment:
if continues to have errors, would never check ctx->Done() which is at the end of the loop. Probably good to either add select to check for ctx->Done() here or move select to top of for loop?